File: binary_search/boj2110.py
import sys

input = sys.stdin.readline

# C개의 공유기를 N개의 집에 인접 거리 D가 최대가 되게 할 때 D
# 이진탐색을 통해 D를 결정 가능 -> res 갱신, M 증가 -> L = M + 1
N, C = map(int, input().split())  # N<2*10^5 C <= N

homes = sorted([int(input()) for _ in range(N)])

left, right = 1, homes[-1]


def possible(d: int, c: int) -> bool:
    # 거리가 d로, c개가 배치가 가능한가?
    cnt = 1  # 첫 위치는 무조건 설치
    curr_idx = 0
    for i in range(1, N):
        if homes[i] - homes[curr_idx] >= d:
            cnt += 1
            curr_idx = i
            if cnt == c:
                return True
    return False


res = 1
while left <= right:
    mid = (left + right) // 2
    if possible(mid, C):
        res = mid
        left = mid + 1
    else:
        right = mid - 1


print(res)

File: java/boj/BOJ2110.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ2110 {

	static FastReader scan = new FastReader();
	static int N, C;
	static int[] homes;

	static void input() {
		N = scan.nextInt();
		C = scan.nextInt();
		homes = new int[N + 1];
		for (int i = 1; i <= N; i++) {
			homes[i] = scan.nextInt();
		}
		Arrays.sort(homes, 1, N);
	}

	static boolean possible(int len) {
		int cnt = 1, last = homes[1];
		for (int i = 2; i <= N; i++) {
			if (homes[i] - last < len)
				continue; // 최소 len만큼 떨어진 곳까지 이동
			last = homes[i];
			cnt++;
		}
		return cnt >= C;
	}

	static void pro(int left, int right) {
		// 1번집은 무조건 설치
		int maxLength = 1; // 최악의 상황 1
		while (left <= right) {
			int mid = (left + right) / 2;
			// 최대거리 mid만큼 떨어진 집에 공유기 설치하고 나머지도 다 설치가능?
			if (possible(mid)) {
				maxLength = mid;
				left = mid + 1;
			} else {
				right = mid - 1;
			}
		}
		System.out.println(maxLength);
	}

	public static void main(String[] args) {
		input();
		pro(1, 1000000000);
	}

	static class FastReader {
		BufferedReader br;
		StringTokenizer st;

		public FastReader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}

		String next() {
			while (st == null || !st.hasMoreElements()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			return st.nextToken();
		}

		int nextInt() {
			return Integer.parseInt(next());
		}

		long nextLong() {
			return Long.parseLong(next());
		}

		String nextLine() {
			String str = "";
			try {
				str = br.readLine();
			} catch (IOException e) {
				e.printStackTrace();
			}

			return str;
		}
	}
}
